auth0/src/main/java/com/auth0/android/request/internal/ThreadSwitcher.kt

```diff
@@ -2,53 +2,125 @@ package com.auth0.android.request.internal
 
 import android.os.Handler
 import android.os.Looper
-import androidx.core.os.HandlerCompat
+import androidx.annotation.VisibleForTesting
+import java.lang.reflect.InvocationTargetException
 import java.util.concurrent.Executor
 import java.util.concurrent.Executors
 
 /**
- * Thread Switcher that makes use of the Main Looper
- * and a background thread Executor.
+ * The maximum concurrent threads to execute
+ * in the background. Value taken from the Android docs.
+ * @see <a href="https://developer.android.com/guide/background/threading#creating-multiple-threads">Android: creating-multiple-threads</a>
  */
-internal object DefaultThreadSwitcher : ThreadSwitcher(
-    Looper.getMainLooper(),
-    Executors.newFixedThreadPool(MAX_CONCURRENT_THREADS)
-)
+private const val MAX_CONCURRENT_THREADS = 4
 
 /**
- * Exposes methods to execute tasks in the background
- * or post tasks in the Main / UI thread.
- * @param mainLooper The Main / UI thread Looper.
+ * Thread Switcher that makes use of the Main Looper
+ * and a background thread Executor.
+ * @param mainHandler The Main / UI thread Looper.
  * @param backgroundExecutor The executor that enqueues tasks to be run in the background.
  */
-public open class ThreadSwitcher(
-    private val mainLooper: Looper,
-    private val backgroundExecutor: Executor
-) {
-    private val mainHandler: Handler by lazy {
-        HandlerCompat.createAsync(mainLooper)
+internal class DefaultThreadSwitcher(
+    private val backgroundExecutor: Executor = Executors.newFixedThreadPool(MAX_CONCURRENT_THREADS)
+) : ThreadSwitcher {
+
+    @Volatile
+    private var mainHandler: Handler? = null
+
+    override fun mainThread(runnable: Runnable) {
+        mainHandler ?: synchronized(this) {
+            if (mainHandler == null) {
+                mainHandler = createAsync(Looper.getMainLooper())
+            }
+        }
+        mainHandler?.post(runnable)
+    }
+
+    override fun backgroundThread(runnable: Runnable) {
+        backgroundExecutor.execute(runnable)
     }
 
-    internal companion object {
-        /**
-         * The maximum concurrent threads to execute
-         * in the background. Value taken from the Android docs.
-         * @see <a href="https://developer.android.com/guide/background/threading#creating-multiple-threads">Android: creating-multiple-threads</a>
-         */
-        internal const val MAX_CONCURRENT_THREADS = 4
+    private fun createAsync(looper: Looper): Handler {
+        try {
+            return Handler::class.java.getDeclaredConstructor(
```

Comment:
😬 using java reflection is a big red flag to us. Is there any other way of achieving this?

Reply:
This is an oversight on my part, I'll try to redo this. Thanks